pynumaflow/sourcer/async_server.py

```diff
@@ -0,0 +1,213 @@
+    ...         yield Message(payload=payload, keys=keys, offset=offset, event_time=event_time)
+    ... async def ack_handler(ack_request: AckRequest):
+    ...     return
+    ... async def pending_handler() -> PendingResponse:
+    ...     PendingResponse(count=10)
+    >>> grpc_server = AsyncSourcer(read_handler=read_handler,
+    ...                     ack_handler=ack_handler,
+    ...                     pending_handler=pending_handler)
+    >>> aiorun.run(grpc_server.start())
+    """
+
+    def __init__(
+        self,
+        read_handler: AsyncSourceReadCallable,
+        ack_handler,
+        pending_handler,
+        sock_path=SOURCE_SOCK_PATH,
+        max_message_size=MAX_MESSAGE_SIZE,
+        max_threads=MAX_THREADS,
+    ):
+        if read_handler is None or ack_handler is None or pending_handler is None:
+            raise ValueError("read_handler, ack_handler and pending_handler are required")
+        self.__source_read_handler: AsyncSourceReadCallable = read_handler
+        self.__source_ack_handler = ack_handler
+        self.__source_pending_handler = pending_handler
+        self.sock_path = f"unix://{sock_path}"
+        self._max_message_size = max_message_size
+        self._max_threads = max_threads
+        self.cleanup_coroutines = []
+        # Collection for storing strong references to all running tasks.
+        # Event loop only keeps a weak reference, which can cause it to
+        # get lost during execution.
+        self.background_tasks = set()
+
+        self._server_options = [
+            ("grpc.max_send_message_length", self._max_message_size),
+            ("grpc.max_receive_message_length", self._max_message_size),
+        ]
+
+    async def ReadFn(
+        self,
+        request: source_pb2.ReadRequest,
+        context: NumaflowServicerContext,
+    ) -> AsyncIterable[source_pb2.ReadResponse]:
+        """
+        Applies a Read function and returns a stream of datum responses.
+        The pascal case function name comes from the proto source_pb2_grpc.py file.
+        """
+
+        async for res in self.__invoke_source_read_stream(
+            Datum(
+                num_records=request.request.num_records,
+                timeout_in_ms=request.request.timeout_in_ms,
+            )
+        ):
+            yield source_pb2.ReadResponse(result=res)
+
+    async def __invoke_source_read_stream(self, req: Datum):
+        try:
+            async for msg in self.__source_read_handler(req):
+                event_time_timestamp = _timestamp_pb2.Timestamp()
+                event_time_timestamp.FromDatetime(dt=msg.event_time)
+                yield source_pb2.ReadResponse.Result(
+                    payload=msg.payload,
+                    keys=msg.keys,
+                    offset=msg.offset.as_dict,
+                    event_time=event_time_timestamp,
+                )
+        except Exception as err:
+            _LOGGER.critical("UDFError, re-raising the error", exc_info=True)
```

Comment:
nit: can be more detailed. `User-Defined Source ReadError`. same for other methods below.